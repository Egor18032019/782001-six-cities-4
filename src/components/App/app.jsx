import React, {PureComponent} from "react";
import PropTypes, {bool} from "prop-types";
import {Switch, Route, Router, Redirect} from "react-router-dom";
import {connect} from "react-redux";
import Notifications, {notify} from 'react-notify-toast';
import Main from "../Main/main.jsx";
import withMain from "../hocs/with-main/with-main.js";
import {getOffersByActiveCity, getDataStatus, getActiveTown, getPlaceCount, getErrorMessage, getFavoritesOffers} from "../../reducer/data/selectors.js";
import {getOffersActive, getCardId} from "../../reducer/offers/selectors.js";
import {getAuthStatus, getEmail, getUsersErrorMessage} from "../../reducer/user/selectors.js";
import {AuthorizationStatus, Operation as UserOperation} from "../../reducer/user/user-reducer.js";
import {Operation as DataOperation} from "../../reducer/data/data-reducer.js";
import history from "../../history";
const MainWrapped = withMain(Main);

import Favorites from "../favorites/favorites.jsx";
import Property from "../property/property.jsx";
import SignIn from "../sign-in/sign-in.jsx";
import {ActionActive} from "../../reducer/offers/offers-reducer.js";
import {ActionTown} from "../../reducer/data/data-reducer.js";
import {AppRoute} from "../../const.js";

class App extends PureComponent {
  constructor(props) {
    super(props);
  }

  _renderApp() {
    const {onMainTitleClick, onCityNameClick, activeTown, placesCount, activeOffers, cardId,
      active, authorizationStatus, email, usersErrorMessage, onFavoriteButtonClick} = this.props;
    // убрал отображение по статусу загрузки и статусу авторизации
    if (active === `mainPages`) {
      return (
        <MainWrapped
          placesCount={placesCount}
          town={activeTown}
          places={activeOffers}
          onMainTitleClick={onMainTitleClick}
          onCityNameClick={onCityNameClick}
          email={email}
          authorizationStatus={authorizationStatus}
          onFavoriteButtonClick={onFavoriteButtonClick}
        />
      );
    } else if (active === `property`) {
      return (
        <Property
          place={activeOffers.find((offer) => {
            return offer.id === cardId;
          })}
          email={email}
          onFavoriteButtonClick={onFavoriteButtonClick}
          authorizationStatus={authorizationStatus}
        />
      );
    } else if (usersErrorMessage && authorizationStatus === AuthorizationStatus.NO_AUTH) {
      let myColor = {background: `#0E1717`, text: `orange`};
      notify.show(`Проверьте введеные данные  ${usersErrorMessage}`, `custom`, 2500, myColor);
    }
    return `что то пошло не так`;
  }

  render() {
    const {onMainTitleClick, onCityNameClick, activeTown, placesCount, activeOffers,
      authorizationStatus, email, onLoginUsers, cardId, errorMessage, usersErrorMessage, onFavoriteButtonClick} = this.props;
    let status = (errorMessage ? notify.show(`${errorMessage}`, `error`) : ``);
    let myColor = {background: `#0E1717`, text: `orange`};
    let statusUser = (usersErrorMessage ? notify.show(`Проверьте введеные данные  ${usersErrorMessage}`, `custom`, 2500, myColor) : ``);
    return (
      <Router
        history={history}>
        <Notifications />
        <Switch>
          <Route exact path={AppRoute.ROOT}>
            {status}
            {this._renderApp()}
          </Route>
          <Route exact path={AppRoute.MAIN}>
            {status}
            <MainWrapped
              placesCount={placesCount}
              town={activeTown}
              places={activeOffers}
              onMainTitleClick={onMainTitleClick}
              onCityNameClick={onCityNameClick}
              email={email}
              authorizationStatus={authorizationStatus}
              onFavoriteButtonClick={onFavoriteButtonClick}
            />
          </Route>
          <Route exact path={AppRoute.PROPERTY}>
            {authorizationStatus === AuthorizationStatus.NO_AUTH ? <Redirect to={AppRoute.LOGIN}/> :
              <Property
                place={activeOffers.find((offer) => {
                  return offer.id === cardId;
                })}
                email={email}
                authorizationStatus={authorizationStatus}
                onFavoriteButtonClick={onFavoriteButtonClick}
              />}
            {statusUser}
          </Route>
          <Route exact path={AppRoute.LOGIN}>
            {status}
            <SignIn
              onLoginUsers={onLoginUsers}
              activeTown={activeTown}
              email={email}
              authorizationStatus={authorizationStatus}
            />
          </Route>
          <Route exact path={AppRoute.FAVORITES}>
            <Favorites
              email={email}
              authorizationStatus={authorizationStatus}
            />
          </Route>

        </Switch>
      </Router >
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  onMainTitleClick(place) {
    dispatch(ActionActive.activeState(place));
  },
  onCityNameClick(city) {
    dispatch(ActionTown.changeCity(city));
  },
  onLoginUsers(authData) {
    dispatch(UserOperation.login(authData));
  },
  onFavoriteButtonClick(place) {
    dispatch(DataOperation.addToFavorite(place));
  }
});

const mapStateToProps = (store) => {
  return ({
    isDataLoaded: getDataStatus(store),
    activeOffers: getOffersByActiveCity(store),
    activeTown: getActiveTown(store),
    placesCount: getPlaceCount(store),
    cardId: getCardId(store),
    active: getOffersActive(store),
    authorizationStatus: getAuthStatus(store),
    email: getEmail(store),
    errorMessage: getErrorMessage(store),
    usersErrorMessage: getUsersErrorMessage(store),
    favoriteOffers: getFavoritesOffers(store),
  });
};

App.propTypes = {
  onCityNameClick: PropTypes.func.isRequired,
  onMainTitleClick: PropTypes.func.isRequired,
  onFavoriteButtonClick: PropTypes.func.isRequired,
  isDataLoaded: bool.isRequired,
  activeTown: PropTypes.string.isRequired,
  placesCount: PropTypes.number.isRequired,
  activeOffers: PropTypes.array.isRequired,
  active: PropTypes.string.isRequired,
  cardId: PropTypes.number,
  authorizationStatus: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  usersErrorMessage: PropTypes.string,
  errorMessage: PropTypes.string,
  onLoginUsers: PropTypes.func.isRequired,
};

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App); // первым стате а вторым диспатчеры


// TODO: Удалите логику рендера компонента «Главная страница»,
// когда по значению аuthorizationStatus равному NO_AUTH рендерится компонент «Sign In».
//  Теперь оба эти компонента должны рендериться только каждый по своему маршруту.
