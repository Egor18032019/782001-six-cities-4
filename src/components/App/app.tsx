import * as React from "react";
import {Switch, Route, Router, Redirect} from "react-router-dom";
import {connect} from "react-redux";
import Main from "../Main/main.jsx";
import withMain from "../hocs/with-main/with-main.js";
import {
  getOffersByActiveCity, getDataStatus, getActiveTown, getPlaceCount, getErrorMessage,
  getFavoritesOffers, getList} from "../../reducer/data/selectors.js";
import {getAuthStatus, getEmail, getUsersErrorMessage} from "../../reducer/user/selectors.js";
import {AuthorizationStatus, Operation as UserOperation} from "../../reducer/user/user-reducer";
import {Operation as DataOperation} from "../../reducer/data/data-reducer";
import {ActionTown} from "../../reducer/data/data-reducer";
import history from "../../history";
const MainWrapped = withMain(Main);

import Favorites from "../favorites/favorites.jsx";
import Property from "../property/property.jsx";
import SignIn from "../sign-in/sign-in.jsx";
import {AppRoute} from "../../const";
import {AppProps} from "../../types";
import withPrivateRoute from "../hocs/with-private-route/with-private-route.js";
const FavoritesPagePrivate = withPrivateRoute(Favorites, AppRoute.LOGIN);

const App = (props:AppProps)=> {

  const {onCityNameClick, activeTown, placesCount, activeOffers,
    authorizationStatus, email, onLoginUsers, cardId,
    onFavoriteButtonClick, cityList, isDataLoaded} = props;
  const isAuthorizationStatus = (authorizationStatus === AuthorizationStatus.AUTH);
  return (
    <Router
      history={history}>
      <Switch>
        <Route exact path={AppRoute.ROOT}>
          <MainWrapped
            placesCount={placesCount}
            town={activeTown}
            cityList={cityList}
            places={activeOffers}
            email={email}
            authorizationStatus={authorizationStatus}
            onCityNameClick={onCityNameClick}
            onFavoriteButtonClick={onFavoriteButtonClick}
          />
        </Route>
        <Route exact path={AppRoute.PROPERTY}
          render={(routeProps) => {
            if (isDataLoaded) {
              return (
                <Property
                  place={cardId}
                  choiseId={routeProps.match.params.id}
                  email={email}
                  authorizationStatus={authorizationStatus}
                  onFavoriteButtonClick={onFavoriteButtonClick}
                />);
            }
            return (`Подождите. Идет загрузка`);
          }}>
        </Route>
        <Route exact path={AppRoute.LOGIN}
          render = {()=>{
            if (isAuthorizationStatus) {
              return <Redirect to={AppRoute.ROOT} />;
            }
            return (<SignIn
              onLoginUsers={onLoginUsers}
              activeTown={activeTown}
              email={email}
              authorizationStatus={authorizationStatus}
            />);
          }}>
        </Route>
        <Route exact path={AppRoute.FAVORITES}>
          <FavoritesPagePrivate
            email={email}
            cityList={cityList}
            authorizationStatus={authorizationStatus}
            onFavoriteButtonClick={onFavoriteButtonClick}
          />
        </Route>
      </Switch>
    </Router >
  );
};


const mapDispatchToProps = (dispatch) => ({
  onCityNameClick(city) {
    dispatch(ActionTown.changeCity(city));
  },
  onLoginUsers(authData) {
    dispatch(UserOperation.login(authData));
  },
  onFavoriteButtonClick(place) {
    dispatch(DataOperation.addToFavorite(place));
  }
});

const mapStateToProps = (store) => {
  return ({
    isDataLoaded: getDataStatus(store),
    activeOffers: getOffersByActiveCity(store),
    activeTown: getActiveTown(store),
    placesCount: getPlaceCount(store),
    authorizationStatus: getAuthStatus(store),
    email: getEmail(store),
    errorMessage: getErrorMessage(store),
    usersErrorMessage: getUsersErrorMessage(store),
    favoriteOffers: getFavoritesOffers(store),
    cityList: getList(store),
  });
};


export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App); // первым стате а вторым диспатчеры


// TODO: Удалите логику рендера компонента «Главная страница»,
// когда по значению аuthorizationStatus равному NO_AUTH рендерится компонент «Sign In».
//  Теперь оба эти компонента должны рендериться только каждый по своему маршруту.
