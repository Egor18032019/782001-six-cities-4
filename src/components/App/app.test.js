import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import App from "./app.jsx";
import NameSpace from "../../reducer/name-space.js";

const mockSettings = [{
  id: 0,
  city: `Amsterdam`,
  type: `Apartament`,
  description: `Beautiful & luxurious apartment at great location`,
  price: 120,
  isBookmark: false,
  isPremium: true,
  rating: 4.8,
  coordinate: [52.3909553943508, 4.85309666406198],
  mainPhoto: `img/apartment-01.jpg`,
  bedrooms: 3,
  maxAdults: 4,
  options: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
  images: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
  stories: [`A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
  `, `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.
  `, `и пусть весь мир подождет..`],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true,
    name: `Angelina`
  }
},
{
  id: 1,
  city: `Amsterdam`,
  type: `Private room`,
  description: `Wood and Stone`,
  price: 80,
  isBookmark: true,
  isPremium: false,
  rating: 3,
  coordinate: [52.369553943508, 4.85309666406198],
  mainPhoto: `img/room.jpg`,
  bedrooms: 0,
  maxAdults: 2,
  options: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
  images: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
  stories: [`И где тут что то будет написано`, `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.
  `, `Быть в Амстердаме и не покурить?`],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true,
    name: `Monro`
  }
},
{
  id: 2,
  city: `Paris`,
  type: `House`,
  description: `big + warm + good`,
  price: 120,
  isBookmark: false,
  isPremium: true,
  rating: 4.8,
  coordinate: [52.3909553943508, 4.929309666406198],
  mainPhoto: `img/apartment-02.jpg`,
  bedrooms: 3,
  maxAdults: 4,
  options: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
  images: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
  stories: [`У нас всегда свежие булочки`, `А из окна площадь Эйфиля видна`, `Лувр где то рядом`],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true,
    name: `Emanuel`
  }
},
{
  id: 4,
  city: `Amsterdam`,
  type: `House`,
  description: `big + warm + good`,
  price: 120,
  isBookmark: false,
  isPremium: true,
  rating: 4.8,
  coordinate: [52.3809553943508, 4.939309666406198],
  mainPhoto: `img/apartment-02.jpg`,
  bedrooms: 3,
  maxAdults: 4,
  options: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
  images: [`img/room.jpg`, `img/apartment-01.jpg`, `img/apartment-02.jpg`, `img/apartment-03.jpg`, `img/studio-01.jpg`, `img/apartment-01.jpg`],
  stories: [`У нас всегда свежие булочки`, `А из окна площадь Эйфиля видна`, `Лувр где то рядом`],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true,
    name: `Emanuel`
  }
}
];

const mockStore = configureStore([]);

describe(`snapshots test App`, () => {
  it(`Render App for good loading`, () => {
    const store = mockStore({
      [NameSpace.DATA]: {
        data: [],
        isDataLoaded: false,
        placesCount: 0,
        town: `Amsterdam`,
        errorMessage: ``
      },
      [NameSpace.OFFERS]: {
        active: `mainPages`,
        cardId: null,
      },
      [NameSpace.USERS]: {
        authorizationStatus: `NO_AUTH`,
        users: ``,
      },
    });

    const tree = renderer
      .create(
          <Provider store={store}>
            < App
              email = {
                `goro5@mail.ru`
              }
              authorizationStatus = {
                `AUTH`
              }
              onMainTitleClick = {
                () => {}
              }
              onCityNameClick = {
                () => {}
              }
              isDataLoaded = {
                true
              }
              cardId = {
                null
              }
              activeTown = {
                `Amsterdam`
              }
              placesCount = {
                121
              }
              activeOffers = {
                mockSettings
              }
              active = {
                `mainPages`
              }
            />
          </Provider>,
          // так как нет контейнера делаем моковый
          {
            createNodeMock: () => document.createElement(`div`)
          })
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it(`Render App for bad loading`, () => {
    const store = mockStore({
      [NameSpace.DATA]: {
        data: [],
        isDataLoaded: false,
        placesCount: 0,
        town: `Amsterdam`,
        errorMessage: ``
      },
      [NameSpace.OFFERS]: {
        active: `mainPages`,
        cardId: null,
      },
      [NameSpace.USERS]: {
        authorizationStatus: `NO_AUTH`,
        users: ``,
      },
    });

    const tree = renderer
      .create(
          <Provider store={store}>
            < App
              email = {
                `goro5@mail.ru`
              }
              authorizationStatus = {
                `AUTH`
              }
              onMainTitleClick = {
                () => {}
              }
              onCityNameClick = {
                () => {}
              }
              isDataLoaded = {
                false
              }
              cardId = {
                null
              }
              activeTown = {
                `Amsterdam`
              }
              placesCount = {
                121
              }
              activeOffers = {
                mockSettings
              }
              active = {
                `mainPages`
              }
            />
          </Provider>,
          // так как нет контейнера делаем моковый
          {
            createNodeMock: () => document.createElement(`div`)
          })
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});


// как настроить Ес линтер что бы он помогал называть  методы и свойтсва компонентов
