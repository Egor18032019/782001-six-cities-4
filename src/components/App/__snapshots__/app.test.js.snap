// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots test App Render App for bad loading 1`] = `
<div
  className="error"
  style={
    Object {
      "color": "red
        ",
      "height": "100%",
      "margin": "auto",
      "paddingTop": "300px",
      "width": "50%",
    }
  }
>
  <p
    className="error__message"
  >
    Ошибка загрузки страницы
  </p>
  <button
    className="error__button"
    onClick={[Function]}
  >
    Попробовать снова
  </button>
</div>
`;

exports[`snapshots test App Render App for good loading 1`] = `
<div
  className="error"
  style={
    Object {
      "color": "red
        ",
      "height": "100%",
      "margin": "auto",
      "paddingTop": "300px",
      "width": "50%",
    }
  }
>
  <p
    className="error__message"
  >
    Ошибка загрузки страницы
  </p>
  <button
    className="error__button"
    onClick={[Function]}
  >
    Попробовать снова
  </button>
</div>
`;
